### ISSUE TITLE
Proposal: Data model

### ISSUE THREAD

**Issue by sbelak** [metabase/metabase#10071]
# Proposal: Data model

Idea behind data models is to be the canonical representation of various types of datasets that: 1) we consider to be best practice, and 2) understand and can be smart about. Users are encouraged to organise their data accordingly, and we'll work on ETL, to transform as many topologies as possible into canonical form. 

They establish an explicit relationship between field types, table types, metrics, and segments expanding on the ideas from #9599, but making the relation explicit rather than arising from the classification process. Bringing in metrics and segments gives us even richer set of semantics as we can now model the notion of which dimensions "go together". In other words, a data model specification can be viewed as a labelled multigraph (ie. multiple edges between same 2 nodes are possible and edges have a "type"). 

Data models should go into `resources/data_models` and be user-extendable (we'll probably need to introduce namespacing/internal folder structure at some point).

## Uses (non-exhaustive)
* Source & target for ETL
* Reference for x-rays (it also removes a lot of the logic from x-rays which would then be relegated mostly to presentation concerns )
* Signal for relevance/ranking/interestingness 

## Example
```
name: Subscriptions
description: Subscription log
refines: Generic Dataset
required_dimensions:
  - StartTime
  - EndTime
  - User
optional_dimensions:
  - Plan
  - Price
  - LTV
  - AccountAge
  - Source
  - Cohort
metrics:
   ActiveSubscriptions: 
     aggregation: [count]
     description: ...
     breakouts:
       - Plan
       - NewUsers
       - Paying
   MRR: 
     aggregation: [sum, [dimension, Price]]
     filter: [>, [dimension, EndTime], [relative-datetime current] ]
     description: Monthly Recurring Revenue
segments:
  NewUsers: 
    filter: [time-interval, [dimension, StartTime], -30, days]
    description: ...
  Paying: 
     filter: [>, [dimension, Price], 0]
     description: ...
  YearlyPlans: 
     filter: [>=, [-, [dimension, EndTime], [dimension, StartTime]]12]
     description: ...
breakout_dimensions: 
  - Plan
  - StartTime
```
All fields except `name` and `required_dimensions` are optional.

In general the model description uses the same conventions as MBTL (see #9847).

`required_dimensions` and `optional_dimensions` refer to fields by their `:special_type`. The assumption/rule being that all dimensions specified in the data models should also be backed by semantic types.

`metrics` and `segments` are not exhaustive, but more of high-information suggestions. 

## Implementation

For a simple round-trip (eg. chrun) a lot of work has already been done as part of MBTL. X-rays code will need some reworking, but mostly throwing away/simplifying both code and heuristics. 

More advanced integration:
* intertwine with type classification
* ETL chains
* Exposing the data-model-as-a-graph (probably via some notion of querying it and/or spiting out properties)

## Open questions:
* Should data models be always confined to a single (probably heavily denormalized) table, or could a single data model stand for a few (presumably interconnected) tables? [_single table is probably enough for now_]

**Comment by sbelak** [comment]
This mostly supersedes the notion of "data fragment" from #9883. The only desirable property that fragments have that this lacks is the notion of grouping dimensions (which would add another layer of semantic information).

**Comment by salsakran** [comment]
I'm a little confused at the presence of "expressions" _and_ metrics. 

Wouldn't it be simpler to allow you to in-line expressions when defining either dimensions or metrics and just stick to those two?

What's the purpose of `breakout_dimensions` ? Is that to say you wouldn't be able to breakout by other dimensions ? Is it intended to be a drill-through path (eg, first click breaks out by `plan` ,second by `StartTime`)? Is this necessary at this stage?

**Comment by sbelak** [comment]
The idea with expressions was to better express relations between fields (eg. duration connects start & end), but it might be too cute (and of little value).

`breakout_dimensions` is more of a hint. My thinking here was 2-fold: making it easier to reason/understand about fat denomralized tables which might have a lot of fields; and to maximize the amount of semantic meaning we can attach to our entities. Again, not strictly necessary, but would certainly help with eg. x-rays.

**Comment by sbelak** [comment]
Each table gets assigned one data model which includes everything we want to do with it (ie. end-user extensibility should wait for v2).

**Comment by sbelak** [comment]
Implemented with #10437
